xdsl/tools/command_line_tool.py

```diff
@@ -4,89 +4,239 @@
 from collections.abc import Callable
 from typing import IO
 
-from xdsl.dialects.affine import Affine
-from xdsl.dialects.arith import Arith
-from xdsl.dialects.builtin import Builtin, ModuleOp
-from xdsl.dialects.cf import Cf
-from xdsl.dialects.cmath import CMath
-from xdsl.dialects.comb import Comb
-from xdsl.dialects.experimental.dmp import DMP
-from xdsl.dialects.experimental.fir import FIR
-from xdsl.dialects.experimental.hls import HLS
-from xdsl.dialects.experimental.math import Math
-from xdsl.dialects.fsm import FSM
-from xdsl.dialects.func import Func
-from xdsl.dialects.gpu import GPU
-from xdsl.dialects.hw import HW
-from xdsl.dialects.irdl.irdl import IRDL
-from xdsl.dialects.linalg import Linalg
-from xdsl.dialects.llvm import LLVM
-from xdsl.dialects.ltl import LTL
-from xdsl.dialects.memref import MemRef
-from xdsl.dialects.mpi import MPI
-from xdsl.dialects.omp import OMP
-from xdsl.dialects.pdl import PDL
-from xdsl.dialects.printf import Printf
-from xdsl.dialects.riscv import RISCV
-from xdsl.dialects.riscv_func import RISCV_Func
-from xdsl.dialects.riscv_scf import RISCV_Scf
-from xdsl.dialects.riscv_snitch import RISCV_Snitch
-from xdsl.dialects.scf import Scf
-from xdsl.dialects.seq import Seq
-from xdsl.dialects.snitch import Snitch
-from xdsl.dialects.snitch_runtime import SnitchRuntime
-from xdsl.dialects.snitch_stream import SnitchStream
-from xdsl.dialects.stencil import Stencil
-from xdsl.dialects.stream import Stream
-from xdsl.dialects.test import Test
-from xdsl.dialects.vector import Vector
-from xdsl.frontend.symref import Symref
+from xdsl.dialects.builtin import ModuleOp
 from xdsl.ir import Dialect, MLContext
 from xdsl.parser import Parser
 from xdsl.passes import ModulePass
 from xdsl.utils.exceptions import ParseError
 
 
-def get_all_dialects() -> list[Dialect]:
+def get_all_dialects() -> list[tuple[str, Callable[[], Dialect]]]:
```

Comment:
Isn't the returned thing here pretty much exactly a `dict[str, Callable[...]]`? Why don't we just make it one?

I think loading lazily in a method called `get_all_dialects()` makes the interface of the cli-tool not quite as intuitive anymore. Couldn't we just add another method? `get_all_dialects_lazy()` or whatever?

Reply:
This is ordered, which I think is a good thing

Reply:
Hmm, I'm wondering if `dict` is indeed better. Right now, we have an order, but it's ordered semi-randomly.
We don't necessarily want people to rely on that order, so I would argue `dict` is better.
And I'm fairly against `get_all_dialects` in a non-lazy way. Having an easy API to load all dialects at once is a bit dangerous because people might use that directly without knowing it's better in a lazy_way.
Also, that's just simplifying the API for something that is likely called only once per project.

Reply:
I guess there exists OrderedDict if we care about the order? Just feels weird to me to see something used like a map/dict but not being called a dict :sweat_smile: 

I guess you are right on the get_all_dialects front :+1:

Reply:
I guess the dict guarantees that the key is unique and that the order is undefined, so it would probably be a slight improvement